lib/vauth/ntlm.c

```diff
@@ -272,7 +275,10 @@ CURLcode Curl_auth_decode_ntlm_type2_message(struct Curl_easy *data,
   unsigned char *type2 = NULL;
   size_t type2_len = 0;
 
-#if defined(USE_NSS)
+#if !defined(USE_OPENSSL) && \
+    !defined(USE_GNUTLS_NETTLE) && \
+    !defined(USE_GNUTLS) && \
+    defined(USE_NSS)
```

Comment:
I am not sure that this is robust.

Granted, you switched the order of the conditionals above, and that happens to favor only OpenSSL and GNU TLS over NSS right now.

But you already have *two* other locations where you have to replicate that order partially: this place here and `lib/http_ntlm.c".

And I bet not even you would remember to update these in, say, 6 months from now, should another GNUTLS variant (or similar) be added.

So I still would prefer the runtime check, as it is a *lot* more robust.

Having said that, I am not a core cURL reviewer, therefore my opinion is a bit less important than e.g. @bagder's.

Reply:
By 'used', I meant 'compiled-in'. Robustness can be argued of course, and most likely it can be made cleaner by moving these checks to a common header or other ways to ensure the correct order.

Reply:
Moving this single check to runtime don't seem correct to me. Moving all checks to runtime seems overkill and bloaty. Though I won't have the resources for a long fight over this, so I'm closing this PR for now. If somebody will still be interested in this, please feel free to use it.

Reply:
Let's revisit this, What about putting something like this in vtls.h
~~~c
/* If NTLM is the first available SSL backend (see order in curl_ntlm_core)
   then it must be initialized to be used by NTLM. */
#if !defined(USE_OPENSSL) && \
    !defined(USE_GNUTLS_NETTLE) && \
    !defined(USE_GNUTLS) && \
    defined(USE_NSS)
#define NTLM_NEEDS_NSS_INIT
#endif
~~~
then the other two files ntlm.c and http_ntlm.c use that define and change the comment block at the beginning to something like
~~~c
/* SSL backend-specific #if branches in this file must be kept in the order
   documented in curl_ntlm_core. */
~~~
this way we get rid of the duplication

Reply:
@jay: It doesn't feel out-of-place this way, indeed, thanks for the suggestion. Added a commit that implements it.